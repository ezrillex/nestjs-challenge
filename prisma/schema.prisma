// This is your Prisma schema file,
// learn more about it in the docs: https://pris.ly/d/prisma-schema

// Looking for ways to speed up your queries, or scale easily with your serverless or edge functions?
// Try Prisma Accelerate: https://pris.ly/cli/accelerate-init

// npx prisma migrate dev --name the_name
generator client {
  provider = "prisma-client-js"
}

datasource db {
  provider = "postgresql"
  url      = env("DATABASE_URL")
}

model Testing {
  id   String @id @default(uuid())
  test String
}

enum roles {
  customer
  manager
  admin
  anonymous // public
}

model Users {
  id                                 String     @id @default(uuid())
  first_name                         String     @db.VarChar(60)
  last_name                          String     @db.VarChar(120)
  email                              String     @unique
  password                           String     @db.Char(60) // hashed and salted password
  created_at                         DateTime   @default(now())
  role                               roles      @default(customer)
  failed_login_attempts              Int        @default(0)
  failed_login_attempts_timestamps   DateTime[] @default([])
  last_login_at                      DateTime?
  password_reset_requests            Int        @default(0)
  password_reset_requests_timestamps DateTime[] @default([])
  password_reset_token               String?
  password_last_updated              DateTime   @default(now())
}
