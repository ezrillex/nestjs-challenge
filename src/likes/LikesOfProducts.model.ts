import { Field, ID, ObjectType } from '@nestjs/graphql';

@ObjectType()
export class LikesOfProducts {
  @Field(() => ID)
  user_id: string;

  @Field(() => ID)
  product_variation_id: string;

  @Field(() => Boolean)
  state: boolean;
}
