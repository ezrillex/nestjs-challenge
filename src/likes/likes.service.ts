import { BadRequestException, Injectable } from '@nestjs/common';
import { PrismaService } from '../prisma/prisma.service';
import { LikesOfProducts } from './LikesOfProducts.model';
import { ProductVariations } from '../products/product_variation/product-variations.model';

@Injectable()
export class LikesService {
  constructor(private readonly prisma: PrismaService) {}

  async toggleLike(
    variation_id: string,
    user_id: string,
  ): Promise<LikesOfProducts> {
    const variation = await this.prisma.productVariations.count({
      where: {
        id: variation_id,
        product: {
          is_published: true,
          is_deleted: false,
        },
      },
    });
    if (variation === 0) {
      throw new BadRequestException('Product Variation not found!');
    }

    const where = {
      user_id: user_id,
      product_variation_id: variation_id,
    };

    const like = await this.prisma.likesOfProducts.count({
      where: where,
    });

    if (like > 0) {
      const data = await this.prisma.likesOfProducts.delete({
        where: {
          user_id_product_variation_id: where,
        },
      });
      return {
        state: false,
        ...data,
      };
    } else {
      const data = await this.prisma.likesOfProducts.create({
        data: {
          user_id: user_id,
          product_variation_id: variation_id,
        },
      });
      return {
        state: true,
        ...data,
      };
    }
  }

  async getLikes(user_id: string): Promise<ProductVariations[]> {
    const likes = await this.prisma.likesOfProducts.findMany({
      where: {
        user_id: user_id,
        likes_product_variation: {
          product: {
            is_published: true,
            is_deleted: false,
          },
        },
      },
      select: {
        likes_product_variation: true,
      },
    });
    return likes.map((like) => like.likes_product_variation);
  }
}
