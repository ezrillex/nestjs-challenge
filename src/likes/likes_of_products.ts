import { Field, ID, ObjectType } from '@nestjs/graphql';
import { Users } from '../auth/users.model';
import { ProductVariations } from '../products/product_variation/product-variations';

@ObjectType()
export class LikesOfProducts {
  @Field(() => ID)
  id: string;

  @Field(() => Users)
  liked_by: Users;

  @Field(() => ProductVariations)
  likes_product_variation: ProductVariations;
}
