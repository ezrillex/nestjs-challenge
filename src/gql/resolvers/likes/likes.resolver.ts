import { Args, Context, Mutation, Query, Resolver } from '@nestjs/graphql';
import { RequiresRole } from '../../../decorators/requires-role/requires-role.decorator';
import { roles } from '@prisma/client';
import { ParseUUIDPipe } from '@nestjs/common';
import { LikesOfProducts } from '../../models/products/likes/likes_of_products/likes_of_products';
import { ProductsService } from '../../../services/products/products.service';

@Resolver()
export class LikesResolver {
  // todo refactor this into its separate service?
  constructor(private productsService: ProductsService) {}

  @RequiresRole(roles.customer)
  @Mutation(() => String, { nullable: true })
  async likeProduct(
    @Args('variation_id', { type: () => String }, ParseUUIDPipe)
    variation_id: string,
    @Context('req') request: Request,
  ) {
    const result = await this.productsService.LikeProduct(
      variation_id,
      request['user'].id,
    );

    return result.id;
  }

  @RequiresRole(roles.customer)
  @Mutation(() => String, { nullable: true })
  async removeLikeProduct(
    @Args('like_id', { type: () => String }, ParseUUIDPipe)
    like_id: string,
  ) {
    return this.productsService.RemoveLike(like_id);
  }

  @RequiresRole(roles.customer)
  @Query(() => [LikesOfProducts], { nullable: true })
  async getLikes(@Context('req') request: Request) {
    return this.productsService.GetLikes(request['user'].id);
  }
}
