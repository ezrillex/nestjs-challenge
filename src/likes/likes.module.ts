import { Module } from '@nestjs/common';
import { LikesService } from './likes.service';
import { LikesResolver } from './likes.resolver';
import { PrismaService } from '../prisma/prisma.service';
import { ProductsService } from '../products/products.service';
import { UsersService } from '../users/users.service';

@Module({
  providers: [
    LikesService,
    LikesResolver,
    PrismaService,
    ProductsService,
    UsersService,
  ],
})
export class LikesModule {}
