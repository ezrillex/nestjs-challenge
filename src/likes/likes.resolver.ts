import {
  Args,
  Context,
  Directive,
  Mutation,
  Parent,
  Query,
  ResolveField,
  Resolver,
} from '@nestjs/graphql';
import { RequiresRole } from '../common/decorators/requires-role.decorator';
import { roles } from '@prisma/client';
import { ParseBoolPipe, ParseUUIDPipe } from '@nestjs/common';
import { LikesService } from './likes.service';
import { ProductsService } from '../products/products.service';
import { UsersService } from '../users/users.service';
import { ProductVariations } from '../products/product_variation/product-variations.model';
import { LikesOfProducts } from './LikesOfProducts.model';

@Resolver()
export class LikesResolver {
  constructor(
    private readonly likesService: LikesService,
    private readonly productsService: ProductsService,
    private readonly usersService: UsersService,
  ) {}

  @Directive('@deprecated(reason: "Use toggleLikes instead")')
  @RequiresRole(roles.customer)
  @Mutation(() => String, { nullable: true })
  async likeProduct(
    @Args('variation_id', { type: () => String }, ParseUUIDPipe)
    variation_id: string,
    @Context('req') request: Request,
  ) {
    return (
      await this.likesService.LikeProduct(variation_id, request['user'].id)
    ).id;
  }

  @Directive('@deprecated(reason: "Use toggleLikes instead")')
  @RequiresRole(roles.customer)
  @Mutation(() => String, { nullable: true })
  async removeLikeProduct(
    @Args('like_id', { type: () => String }, ParseUUIDPipe)
    like_id: string,
  ) {
    return (await this.likesService.RemoveLike(like_id)).result;
  }

  @RequiresRole(roles.customer)
  @Mutation(() => LikesOfProducts)
  async toggleLike(
    @Args('product_variation_id', { type: () => String }, ParseUUIDPipe)
    id: string,
    @Args('status', { type: () => Boolean }, ParseBoolPipe)
    status: boolean,
    @Context('req') request: Request,
  ) {
    return this.likesService.ToggleLike(id, status, request['user'].id);
  }

  @RequiresRole(roles.customer)
  @Query(() => [ProductVariations], { nullable: true })
  async getLikes(@Context('req') request: Request) {
    return this.likesService.GetLikes(request['user'].id);
  }
}
