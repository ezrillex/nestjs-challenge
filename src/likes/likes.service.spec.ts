import { Test, TestingModule } from '@nestjs/testing';
import { LikesService } from './likes.service';
import { PrismaService } from '../prisma/prisma.service';

describe('LikesService', () => {
  let service: LikesService;
  let prismaService: PrismaService;

  beforeEach(async () => {
    const module: TestingModule = await Test.createTestingModule({
      providers: [PrismaService, LikesService],
    }).compile();

    service = module.get<LikesService>(LikesService);
    prismaService = module.get<PrismaService>(PrismaService);
  });

  it('should be defined', () => {
    expect(service).toBeDefined();
  });

  it('should have method ToggleLike', () => {
    expect(service.ToggleLike).toBeDefined();
  });

  it('should have method Get Likes of User', () => {
    expect(service.GetLikes).toBeDefined();
  });

  describe('toggle likes', () => {
    it('throws if variation to like is invalid', async () => {
      const spy = jest.spyOn(prismaService.productVariations, 'count');
      await expect(
        service.ToggleLike(
          'c0177e32-5584-4518-bbe8-5a648fa33f85',
          'c0177e32-5584-4518-bbe8-5a648fa33f85',
        ),
      ).rejects.toThrowErrorMatchingSnapshot('prod var not found');
      expect(spy).toHaveBeenCalledWith({
        where: { id: 'c0177e32-5584-4518-bbe8-5a648fa33f85' },
      });
    });
    it('if the like exists it should delete the database record', async () => {
      const ids = {
        user_id: 'c0177e32-5584-4518-bbe8-5a648fa33f85',
        product_variation_id: 'c0177e32-5584-4518-bbe8-5a648fa33f85',
      };
      const spy = jest
        .spyOn(prismaService.productVariations, 'count')
        .mockResolvedValue(1);
      const likeSpy = jest
        .spyOn(prismaService.likesOfProducts, 'count')
        .mockResolvedValue(1);
      const deleteResult = {
        id: 'c0177e32-5584-4518-bbe8-5a648fa33f85',
        ...ids,
        created_at: new Date(2020, 12, 12, 12, 12, 12),
      };
      const deleteSpy = jest
        .spyOn(prismaService.likesOfProducts, 'delete')
        .mockResolvedValue(deleteResult);
      await expect(
        service.ToggleLike(
          'c0177e32-5584-4518-bbe8-5a648fa33f85',
          'c0177e32-5584-4518-bbe8-5a648fa33f85',
        ),
      ).resolves.toEqual({
        state: false,
        ...deleteResult,
      });
      expect(spy).toHaveBeenCalledWith({
        where: { id: 'c0177e32-5584-4518-bbe8-5a648fa33f85' },
      });
      expect(likeSpy).toHaveBeenCalledWith({
        where: ids,
      });
      expect(deleteSpy).toHaveBeenCalledWith({
        where: {
          user_id_product_variation_id: ids,
        },
      });
    });
    it('if the like does not exist it should create a database record', async () => {
      const ids = {
        user_id: 'c0177e32-5584-4518-bbe8-5a648fa33f85',
        product_variation_id: 'c0177e32-5584-4518-bbe8-5a648fa33f85',
      };
      const spy = jest
        .spyOn(prismaService.productVariations, 'count')
        .mockResolvedValue(1);
      const likeSpy = jest
        .spyOn(prismaService.likesOfProducts, 'count')
        .mockResolvedValue(0);
      const createResult = {
        id: 'c0177e32-5584-4518-bbe8-5a648fa33f85',
        ...ids,
        created_at: new Date(2020, 12, 12, 12, 12, 12),
      };
      const createSpy = jest
        .spyOn(prismaService.likesOfProducts, 'create')
        .mockResolvedValue(createResult);
      await expect(
        service.ToggleLike(
          'c0177e32-5584-4518-bbe8-5a648fa33f85',
          'c0177e32-5584-4518-bbe8-5a648fa33f85',
        ),
      ).resolves.toEqual({
        state: true,
        ...createResult,
      });
      expect(spy).toHaveBeenCalledWith({
        where: { id: 'c0177e32-5584-4518-bbe8-5a648fa33f85' },
      });
      expect(likeSpy).toHaveBeenCalledWith({
        where: ids,
      });
      expect(createSpy).toHaveBeenCalledWith({
        data: ids,
      });
    });
  });

  describe('get likes', () => {
    it('should pass a query to prisma', async () => {
      const spy = jest.spyOn(prismaService.likesOfProducts, 'findMany');

      await expect(
        service.GetLikes('8b3ae683-0626-44be-b591-9271e288388f'),
      ).resolves.toEqual([]);
      expect(spy.mock.calls).toMatchSnapshot('prisma query get likes');
    });
  });
});
