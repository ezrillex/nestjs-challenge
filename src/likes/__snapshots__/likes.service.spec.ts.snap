// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LikesService get likes should pass a query to prisma: prisma query get likes 1`] = `
[
  [
    {
      "select": {
        "likes_product_variation": true,
      },
      "where": {
        "user_id": "8b3ae683-0626-44be-b591-9271e288388f",
      },
    },
  ],
]
`;

exports[`LikesService toggle likes throws if variation to like is invalid: prod var not found 1`] = `"Product Variation not found!"`;
